### PR TITLE
Add platform flag to template store

### LINKED ISSUE #570: Filter templates by platform
We can filter the templates by platform for the times user is interested in writing a function on `armhf` devidce

## Expected Behaviour

Filter templates by `PLATFORM`

## Current Behaviour

We cannot filter templates  by the platform we are interested in

## Possible Solution

We can add two additional flags:

* One for the user to check what platforms are available. Example: `faas-cli template store list --platforms`
* One to filter the actual output of the `template store list`. Example `faas-cli template store list --filter armhf`

## Steps to Reproduce (for bugs)

N.A. it is a feature

## Context

N.A. just sometimes you would like to see the platform you are interested in

### PR DESCRIPTION
**Pull request by martindekov**
Adding platform flag to template store to filter
the templates by platform

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

Adding `--platform` flag to filter by platform

## Description

Adding `--platform`/`-p` flag to filter templates by existing platforms from `PLATFORM` field 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Closes #570 

## How Has This Been Tested?

Manually:

Normal invocation of command:
```
> ./faas-cli template store list

NAME                    SOURCE             DESCRIPTION
csharp                  openfaas           Official C# template
dockerfile              openfaas           Official Dockerfile template
go-armhf                openfaas           Official Golang armhf template
go                      openfaas           Official Golang template
java8                   openfaas           Official Java 8 template
node-arm64              openfaas           Official NodeJS 8 arm64 template
node                    openfaas           Official NodeJS 8 template
php7                    openfaas           Official PHP 7 template
python-armhf            openfaas           Official Python 2.7 armhf template
python                  openfaas           Official Python 2.7 template
python3-armhf           openfaas           Official Python 3.6 armhf template
python3                 openfaas           Official Python 3.6 template
ruby                    openfaas           Official Ruby 2.5 template
node10-express-armhf    openfaas-incubator NodeJS 10 Express armhf template
node10-express          openfaas-incubator NodeJS 10 Express template
ruby-http               openfaas-incubator Ruby 2.4 HTTP template
python27-flask          openfaas-incubator Python 2.7 Flask template
python3-flask-armhf     openfaas-incubator Python 3.6 Flask armhf template
python3-flask           openfaas-incubator Python 3.6 Flask template
node8-express-armhf     openfaas-incubator NodeJS 8 Express armhf template
node8-express           openfaas-incubator NodeJS 8 Express template
golang-http-armhf       openfaas-incubator Golang HTTP armhf template
golang-http             openfaas-incubator Golang HTTP template
golang-middleware-armhf openfaas-incubator Golang Middleware armhf template
golang-middleware       openfaas-incubator Golang Middleware template
python3-debian          openfaas-incubator Python 3.6 Debian template
powershell-template     openfaas-incubator Powershell Ubuntu:16.04 template
rust                    booyaa             Rust template
crystal                 tpei               Crystal template
```
Invocation with all available platforms:
```
> ./faas-cli template store list --platform armhf

NAME                    SOURCE             DESCRIPTION
go-armhf                openfaas           Official Golang armhf template
python-armhf            openfaas           Official Python 2.7 armhf template
python3-armhf           openfaas           Official Python 3.6 armhf template
node10-express-armhf    openfaas-incubator NodeJS 10 Express armhf template
python3-flask-armhf     openfaas-incubator Python 3.6 Flask armhf template
node8-express-armhf     openfaas-incubator NodeJS 8 Express armhf template
golang-http-armhf       openfaas-incubator Golang HTTP armhf template
golang-middleware-armhf openfaas-incubator Golang Middleware armhf template

 > ./faas-cli template store list --platform arm64

NAME       SOURCE   DESCRIPTION
node-arm64 openfaas Official NodeJS 8 arm64 template

> ./faas-cli template store list --platform x86_64

NAME                SOURCE             DESCRIPTION
csharp              openfaas           Official C# template
dockerfile          openfaas           Official Dockerfile template
go                  openfaas           Official Golang template
java8               openfaas           Official Java 8 template
node                openfaas           Official NodeJS 8 template
php7                openfaas           Official PHP 7 template
python              openfaas           Official Python 2.7 template
python3             openfaas           Official Python 3.6 template
ruby                openfaas           Official Ruby 2.5 template
node10-express      openfaas-incubator NodeJS 10 Express template
ruby-http           openfaas-incubator Ruby 2.4 HTTP template
python27-flask      openfaas-incubator Python 2.7 Flask template
python3-flask       openfaas-incubator Python 3.6 Flask template
node8-express       openfaas-incubator NodeJS 8 Express template
golang-http         openfaas-incubator Golang HTTP template
golang-middleware   openfaas-incubator Golang Middleware template
python3-debian      openfaas-incubator Python 3.6 Debian template
powershell-template openfaas-incubator Powershell Ubuntu:16.04 template
rust                booyaa             Rust template
crystal             tpei               Crystal template

```

Command with wrong string aka not supported:
```
> ./faas-cli template store list --platform randomstring

Currently supported platforms are: armhf, arm64 and x86_64. Unable to find: randomstring

```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Derek set reviewer: rgee0

**Comment by martindekov** [comment]
Aaaah thanks @rgee0 for pointing that out I will fix the issue and the PR

**Comment by martindekov** [comment]
I will propose how to do it in the comments on the issue so you can give me direction which is better, can we do that ?

**Comment by martindekov** [comment]
https://github.com/openfaas/faas-cli/issues/570

**Comment by martindekov** [comment]
@rgee0 I updated the PR, now I think it does what it should, thanks for letting me know about the wrong context

File: commands/template_store_list.go [hunk 1]
```diff
@@ -117,25 +129,51 @@ func formatTemplatesOutput(templates []TemplateInfo, verbose bool) string {
...
-	fmt.Fprintf(lineWriter, "NAME\tSOURCE\tDESCRIPTION\n")
-	for _, template := range templates {
-		fmt.Fprintf(lineWriter, "%s\t%s\t%s\n",
-			template.TemplateName,
-			template.Source,
-			template.Description)
+func formatBasicOutput(lineWriter *tabwriter.Writer, templates []TemplateInfo, platform string) {
+	if platform != allPlatforms {
+		fmt.Fprintf(lineWriter, "NAME\tSOURCE\tDESCRIPTION\n")
+		for _, template := range templates {
```
Comment (rgee0) [inline commands/template_store_list.go]: Feels like we should avoid repeated code here by pre-filtering the templates of interest.  So we lift out the two checks and then just output what remains.
```
if platform != allPlatforms {
templates = filterTemplates(platform)
}

fmt.Fprintf(lineWriter, "NAME\tLANGUAGE\tPLATFORM\tSOURCE\tDESCRIPTION\n")
for _, template := range templates {
fmt.Fprintf(lineWriter, "%s\t%s\t%s\t%s\t%s\n",
    template.TemplateName,
    template.Language,
    template.Platform,
    template.Source,
    template.Description)
}
```
WDYT?

Comment (rgee0) [reply commands/template_store_list.go]: Depending on what we choose in the check function we might need to handle case in filterTemplates

File: commands/template_store_list.go [hunk 2]
```diff
@@ -149,3 +187,15 @@ type TemplateInfo struct {
 	Repository   string `json:"repo"`
 	Official     string `json:"official"`
 }
+
+func checkExistingPlatforms(platform string) error {
```
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: 1. Yeah it does the same but was unsure if it is a good practice to directly return. I will re-do it to directly return nil or error. 
2. Totally, haven't though of case sensitivity, it is better approach, will re-do it

[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Yeah of course, I will also change the name as proposed

File: commands/template_store_list.go [hunk 3]
```diff
@@ -117,25 +129,51 @@ func formatTemplatesOutput(templates []TemplateInfo, verbose bool) string {
...
-	fmt.Fprintf(lineWriter, "NAME\tLANGUAGE\tPLATFORM\tSOURCE\tDESCRIPTION\n")
-	for _, template := range templates {
-		fmt.Fprintf(lineWriter, "%s\t%s\t%s\t%s\t%s\n",
-			template.TemplateName,
-			template.Language,
-			template.Platform,
-			template.Source,
-			template.Description)
+func formatVerboseOutput(lineWriter *tabwriter.Writer, templates []TemplateInfo, platform string) {
+	if platform != allPlatforms {
```
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Yeah good idea I will implement `filterTemplates`

File: commands/template_store_list.go (hunk 1 above)
Comment (martindekov) [reply commands/template_store_list.go]: I think this is a great proposal I will implement `filterTemplates` to modify the object to contain only the requested platform if one is requested I will also add tests to the two added functions

**Comment by martindekov** [comment]
Thank you for the review @rgee0 I will implement the additional methods tweak the naming and add some tests to showscase how the additional methods work

File: commands/template_store_list.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Now return directly the error or nil ✔️

File: commands/template_store_list.go (hunk 1 above)
Comment (martindekov) [reply commands/template_store_list.go]: Implemented `filterTempaltes` and removed code duplication ✔️

File: commands/template_store_list.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: `filterTemplates` implemented ✔️

File: commands/template_store_list.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Added tests for the `filterTemplates()` and `checkAvailableTempaltes()` functions ✔️

**Comment by martindekov** [comment]
Thank you again @rgee0 I had some job today that's why I could not reference your comments but did it as soon as I could 🍻

File: commands/template_store_list.go [hunk 4]
```diff
@@ -149,3 +187,15 @@ type TemplateInfo struct {
...
 	Official     string `json:"official"`
 }
+
+func checkExistingPlatforms(platform string) error {
+	var err error
+	for lastPlatform, availablePlatform := range availablePlatforms {
+		if availablePlatform == platform {
+			break
+		} else if len(availablePlatforms)-1 == lastPlatform {
+			err = fmt.Errorf("\nCurrently supported platforms are: armhf, arm64 and x86_64. Unable to find: %s\n\n", platform)
```
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Ok I will remove this logic

File: commands/template_store_list_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+func Test_CheckAvailablePlatforms_ValidInput(t *testing.T) {
+	validPlatforms := []string{"ARMHF", "armhf", "ARM64", "arm64", "X86_64", "x86_64"}
+	for _, validPlatform := range validPlatforms {
+		err := checkAvailablePlatforms(validPlatform)
+		if err != nil {
+			t.Errorf("Unexpected error: %s expected: nil", err.Error())
+		}
+	}
+}
+func Test_CheckAvailablePlatforms_InvalidInput(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list_test.go]: Will remove that

File: commands/template_store_list.go [hunk 6]
```diff
@@ -20,15 +20,19 @@ import (
...
 	// DefaultTemplatesStore is the URL where the official store can be found
 	DefaultTemplatesStore = "https://raw.githubusercontent.com/openfaas/store/master/templates.json"
+	allPlatforms          = "allPlatforms"
 )
 
 var (
-	templateStoreURL string
+	templateStoreURL   string
+	platform           string
+	availablePlatforms = [...]string{"armhf", "x86_64", "arm64"}
```
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: I will string-match

**Comment by martindekov** [comment]
Ok @alexellis thanks for the review I will try to do it by EOD

File: commands/template_store_list.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list.go]: Now I string match

File: commands/template_store_list_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply commands/template_store_list_test.go]: removed tests associated with `checkAvailablePlatform()` function and the function itself ✔️

**Comment by martindekov** [comment]
I removed `checkAvailablePlatforms()` function and associated tests and made string match which returns 
```
> ./faas-cli templatestore list --platform asd

NAME SOURCE DESCRIPTION

```
In case a platform does not exist in the example `asd`

**Comment by alexellis** [comment]
> Normal invocation of command:

This should filter all but `x86_64` by default.

**Comment by martindekov** [comment]
```
> ./faas-cli template store list -v

NAME                LANGUAGE   PLATFORM SOURCE             DESCRIPTION
csharp              C#         x86_64   openfaas           Official C# template
dockerfile          Dockerfile x86_64   openfaas           Official Dockerfile template
go                  Go         x86_64   openfaas           Official Golang template
java8               Java       x86_64   openfaas           Official Java 8 template
node                NodeJS     x86_64   openfaas           Official NodeJS 8 template
php7                PHP        x86_64   openfaas           Official PHP 7 template
python              Python     x86_64   openfaas           Official Python 2.7 template
python3             Python     x86_64   openfaas           Official Python 3.6 template
ruby                Ruby       x86_64   openfaas           Official Ruby 2.5 template
node10-express      NodeJS     x86_64   openfaas-incubator NodeJS 10 Express template
ruby-http           Ruby       x86_64   openfaas-incubator Ruby 2.4 HTTP template
python27-flask      Python     x86_64   openfaas-incubator Python 2.7 Flask template
python3-flask       Python     x86_64   openfaas-incubator Python 3.6 Flask template
node8-express       NodeJS     x86_64   openfaas-incubator NodeJS 8 Express template
golang-http         Go         x86_64   openfaas-incubator Golang HTTP template
golang-middleware   Go         x86_64   openfaas-incubator Golang Middleware template
python3-debian      Python     x86_64   openfaas-incubator Python 3.6 Debian template
powershell-template Python     x86_64   openfaas-incubator Powershell Ubuntu:16.04 template
rust                Rust       x86_64   booyaa             Rust template
crystal             Crystal    x86_64   tpei               Crystal template
```
now `x86_64` is the default I hope that's what you meant ^

**Comment by alexellis** [comment]
LGTM 👍

**Pull request merged**